tests/test_hawkdove.py

```diff
@@ -31,66 +41,78 @@ def test_agent_initial_choice():
+        grid_size, include_diagonals=False, hawk_odds=0.3, agent_risk_level=2
+    )
     initial_choices = [a.choice for a in model.schedule.agents]
     choice_count = Counter(initial_choices)
     # expect about 30% hawks
     expected_hawks = model.num_agents * 0.3
     assert math.isclose(choice_count[Play.HAWK], expected_hawks, rel_tol=0.05)
 
 
+def test_base_agent_risk_level():
+    # base class should raise error because method to set risk level is not defined
+    with pytest.raises(NotImplementedError):
+        HawkDoveAgent(1, Mock())
+
+
 def test_agent_initial_risk_level():
-    agent = HawkDoveAgent(1, Mock(), risk_level=2)
+    # single risk agent sets risk level based on model
+    agent = HawkDoveSingleRiskAgent(1, Mock(agent_risk_level=2))
     assert agent.risk_level == 2
 
 
 def test_agent_repr():
     agent_id = 1
     risk_level = 3
-    agent = HawkDoveAgent(agent_id, Mock(), risk_level=risk_level)
-    assert repr(agent) == f"<HawkDoveAgent id={agent_id} r={risk_level}>"
+    agent = HawkDoveSingleRiskAgent(agent_id, Mock(agent_risk_level=risk_level))
+    assert repr(agent) == f"<HawkDoveSingleRiskAgent id={agent_id} r={risk_level}>"
 
 
 def test_model_single_risk_level():
     risk_level = 3
-    model = HawkDoveModel(
-        5, include_diagonals=True, risk_attitudes="single", agent_risk_level=risk_level
+    model = HawkDoveSingleRiskModel(
+        5, include_diagonals=True, agent_risk_level=risk_level
     )
     for agent in model.schedule.agents:
         assert agent.risk_level == risk_level
 
     # handle zero properly (should not be treated the same as None)
     risk_level = 0
-    model = HawkDoveModel(
-        5, include_diagonals=True, risk_attitudes="single", agent_risk_level=risk_level
+    model = HawkDoveSingleRiskModel(
+        5, include_diagonals=True, agent_risk_level=risk_level
     )
     for agent in model.schedule.agents:
         assert agent.risk_level == risk_level
 
 
-def test_model_variable_risk_level():
-    model = HawkDoveModel(5, include_diagonals=True, risk_attitudes="variable")
+def test_variable_risk_level():
+    model = HawkDoveVariableRiskModel(
+        5,
+        include_diagonals=True,
+    )
     # when risk level is variable/random, agents should have different risk levels
     risk_levels = set([agent.risk_level for agent in model.schedule.agents])
     assert len(risk_levels) > 1
 
 
 def test_num_dove_neighbors():
     # initialize an agent with a mock model
-    agent = HawkDoveAgent(1, Mock())
+    agent = HawkDoveSingleRiskAgent(1, Mock(agent_risk_level=2))
     mock_neighbors = [
         Mock(last_choice=Play.HAWK),
         Mock(last_choice=Play.HAWK),
         Mock(last_choice=Play.HAWK),
         Mock(last_choice=Play.DOVE),
     ]
 
-    with patch.object(HawkDoveAgent, "neighbors", mock_neighbors):
+    with patch.object(HawkDoveSingleRiskAgent, "neighbors", mock_neighbors):
         assert agent.num_dove_neighbors == 1
```

Comment:
These are some really nice tests. Are any of them 'integration' tests would you say (making sure results match expectation from theory/math, e.g.) or are these mainly unit tests (making sure code does what it says it does)?

Reply:
I think of integration tests as testing interactions between different systems, like tests in some of our other apps that check that solr results return what we expect for a given set of test input in the solr index. I don't think I've written anything here that isn't a unit test, but the ones that have probabilistic results are a little weird and more mathy, maybe...